src/plugins/gitresolver/gitresolver.c

```diff
@@ -0,0 +1,512 @@
+	elektraFree (objID);
+#endif
+	if (data->tracking == OBJECT)
+	{
+		char * newObj = hasNewObjectCommit (data, blob);
+		if (!newObj)
+		{
+#ifdef DEVMODE
+			fprintf (stderr, "Get: no new object commits - nothing to do\n");
+#endif
+			git_object_free (blob);
+			git_repository_free (repo);
+			git_libgit2_shutdown ();
+			return 0;
+		}
+		else
+		{
+			if (data->objID)
+			{
+				elektraFree (data->objID);
+				data->objID = newObj;
+			}
+			else
+			{
+				data->objID = newObj;
+			}
+		}
+		elektraPluginSetData (handle, data);
+		data = elektraPluginGetData (handle);
+	}
+	FILE * outFile;
+	outFile = fopen (data->tmpFile, "w+");
+	if (!outFile)
+	{
+#ifdef DEVMODE
+		fprintf (stderr, "Couldn't open File %s\n", data->tmpFile);
+#endif
+		ELEKTRA_SET_ERRORF (ELEKTRA_ERROR_COULD_NOT_OPEN, parentKey, "Failed to check out file %s to %s\n", data->file,
+				    data->tmpFile);
+		git_object_free (blob);
+		git_repository_free (repo);
+		git_libgit2_shutdown ();
+		return -1;
+	}
+	fwrite (git_blob_rawcontent ((git_blob *)blob), (size_t)git_blob_rawsize ((git_blob *)blob), 1, outFile);
+	fclose (outFile);
+	git_object_free (blob);
+	git_repository_free (repo);
+	git_libgit2_shutdown ();
+	return 1; // success
+}
+
+
+int elektraGitresolverSet (Plugin * handle ELEKTRA_UNUSED, KeySet * returned ELEKTRA_UNUSED, Key * parentKey ELEKTRA_UNUSED)
+{
+	// get all keys
+	// this function is optional
+	GitData * data = elektraPluginGetData (handle);
+	if (!data) return -1;
+	keySetString (parentKey, data->tmpFile);
+	git_repository * repo = connectToLocalRepo (data);
+	if (!repo)
+	{
+		ELEKTRA_SET_ERRORF (147, parentKey, "Failed to open Repository %s\n", data->repo);
+		git_libgit2_shutdown ();
+		return -1;
+	}
+	const git_oid * headObj = getHeadRef (data, repo);
+	if (!headObj)
+	{
+		ELEKTRA_SET_ERRORF (147, parentKey, "Failed to get reference %s\n", data->refName);
+		git_repository_free (repo);
+		git_libgit2_shutdown ();
+		return -1;
+	}
+	if (data->tracking == HEAD)
+	{
+		char * newCommit = hasNewCommit (data, headObj);
+		if (newCommit)
+		{
+			// newer commit in repo - abort
+			ELEKTRA_SET_ERROR (148, parentKey, "The repository has been updated and is ahead of you");
+			elektraFree (newCommit);
+			git_repository_free (repo);
+			git_libgit2_shutdown ();
+			return -1;
+		}
+		elektraFree (newCommit);
+	}
+	if (data->tracking == OBJECT)
+	{
+		git_object * blob = getBlob (data, repo);
+		if (blob)
+		{
+			char * newObj = hasNewObjectCommit (data, blob);
+			if (newObj)
+			{
+				ELEKTRA_SET_ERROR (148, parentKey, "The repository has been updated and is ahead of you");
+				elektraFree (newObj);
+				git_object_free (blob);
+				git_repository_free (repo);
+				git_libgit2_shutdown ();
+				return -1;
+			}
+			git_object_free (blob);
+		}
+	}
+	if (!data->setPhase)
+	{
+#ifdef DEVMODE
+		fprintf (stderr, "setresolver\n");
+#endif
+		++(data->setPhase);
+	}
+	else if (data->setPhase == 1)
+	{
+// commit phase
```

Comment:
i have to rewrite that part anyway, it will be split into several functions. 
i'm thinking about specifying commit messages as plugin configuration with placeholders for at least filename and time